### PR TITLE
Add support for AbortSignal in addEventListener

### LINKED ISSUE #3327: Feature request: add AbortController support to addEventListener
It is possible to provide in a `signal` (`AbortSignal`) as a part of the `AddEventListenerOptions` object. When this `signal` is aborted, any event listeners that were provided with that `signal` will be unbound.

- [`AbortSignal`](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal)
- [spec](https://dom.spec.whatwg.org/#ref-for-dom-eventtarget-addeventlistener%E2%91%A2)
- [Browser support](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#browser_compatibility) - most noteably using `signal` is currently not supported in Safari

Given there is not universal support yet, should this be added to jsdom?

I was thinking of having a crack at adding `AbortSignal` support to `EventTarget-impl.js`

### PR DESCRIPTION
**Pull request by cheap-glitch**
Resolves #3327

### CODE REVIEW

File: lib/jsdom/living/events/EventTarget-impl.js [hunk 1]
```diff
@@ -27,6 +27,10 @@ class EventTargetImpl {
   addEventListener(type, callback, options) {
     options = normalizeEventHandlerOptions(options, ["capture", "once", "passive"]);
 
+    if (options.signal !== null && options.signal.aborted) {
+      return;
+    }
```
Comment (cheap-glitch) [inline lib/jsdom/living/events/EventTarget-impl.js:32]: https://dom.spec.whatwg.org/#add-an-event-listener

> 2. If listener’s [signal](https://dom.spec.whatwg.org/#event-listener-signal) is not null and is [aborted](https://dom.spec.whatwg.org/#abortsignal-aborted), then return.

File: lib/jsdom/living/events/EventTarget-impl.js [hunk 2]
```diff
@@ -49,6 +53,12 @@ class EventTargetImpl {
       callback,
       options
     });
+
+    if (options.signal !== null) {
+      options.signal._addAlgorithm(() => {
+        this.removeEventListener(type, callback, options);
+      });
+    }
```
Comment (cheap-glitch) [inline lib/jsdom/living/events/EventTarget-impl.js:61]: https://dom.spec.whatwg.org/#add-an-event-listener

> 5. If listener’s [signal](https://dom.spec.whatwg.org/#event-listener-signal) is not null, then [add the following](https://dom.spec.whatwg.org/#abortsignal-add) abort steps to it:
>    1. [Remove an event listener](https://dom.spec.whatwg.org/#remove-an-event-listener) with eventTarget and listener.

File: lib/jsdom/living/events/EventTarget-impl.js [hunk 3]
```diff
@@ -366,7 +376,7 @@ function innerInvokeEventListeners(eventImpl, listeners, phase, itemInShadowTree
  * @returns {Object} object containing at least the "defaultBoolKeys"
  */
 function normalizeEventHandlerOptions(options, defaultBoolKeys) {
-  const returnValue = {};
+  const returnValue = { signal: null };
```
Comment (cheap-glitch) [inline lib/jsdom/living/events/EventTarget-impl.js:379]: Not sure about this, but it helps adhering to the spec in `addEventListener`

**Comment by cheap-glitch** [comment]
Failing test is unrelated to this PR, it fails on `master` too

**Comment by fregante** [comment]
@domenic is there anything else to do here?

**Comment by domenic** [comment]
No, I just haven't had time to review on recent weekends, and this is stuck behind other higher-priority items such as fixing jsdom on Node v18 where it is currently totally broken (see #3358).

**Review by domenic — APPROVED** [review]

**Pull request merged**